Implement EnhancedSecurity HTTP and heuristics tracking.

ronan-apple opened the pull request:
#### 93ffc509729c05e6a7c9920d1fe59d90fa092458
<pre>
Implement EnhancedSecurity HTTP and heuristics tracking.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302003">https://bugs.webkit.org/show_bug.cgi?id=302003</a>
<a href="https://rdar.apple.com/159219681">rdar://159219681</a>

Reviewed by NOBODY (OOPS!).

This change enables usage of EnhancedSecurity when various conditions occur.

In particular, if an insecure HTTP request is made for the main frame, which
is not later upgraded to a HTTPS site, we will enable EnhancedSecurity.

Various heuristics then apply to determine whether we will retain this
EnhancedSecurity state for subsequent navigations, which this change implements.

When a UI initiated navigation occurs we will disable EnhancedSecurity, dropping
back to a regular tab state.

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::hasLocalStorageOrCookies):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/EnhancedSecurity.h: Added.
(WebKit::enhancedSecurityEnabledForState):
(WebKit::enhancedSecurityStatesAreConsistent):
* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::setEnhancedSecurity):
(WebKit::WebBackForwardListItem::enhancedSecurity const):
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
(WebKit::toString):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::setCurrentRequest):
(API::Navigation::setSiteHasStorage):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::currentRequestSiteHasStorage const):
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _allWebsiteDataTypesIncludingPrivate]):
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp: Added.
(WebKit::enabledSitesMap):
(WebKit::didSitePreviouslyUseEnhancedSecurity):
(WebKit::EnhancedSecurityTracking::initializeWithWebsiteDataStore):
(WebKit::EnhancedSecurityTracking::updateEnhancedSecurityDomains):
(WebKit::EnhancedSecurityTracking::transfer):
(WebKit::EnhancedSecurityTracking::enhancedSecurityState const):
(WebKit::EnhancedSecurityTracking::reset):
(WebKit::EnhancedSecurityTracking::makeDormant):
(WebKit::EnhancedSecurityTracking::makeActive):
(WebKit::reasonForEnhancedSecurity):
(WebKit::EnhancedSecurityTracking::enableFor):
(WebKit::EnhancedSecurityTracking::trackChangingSiteNavigation):
(WebKit::EnhancedSecurityTracking::trackSameSiteNavigation):
(WebKit::EnhancedSecurityTracking::enableIfRequired):
(WebKit::EnhancedSecurityTracking::handleBackForwardNavigation):
(WebKit::EnhancedSecurityTracking::trackNavigation):
* Source/WebKit/UIProcess/EnhancedSecurityTracking.h: Added.
* Source/WebKit/UIProcess/Media/RemoteMediaSessionManagerProxy.cpp:
* Source/WebKit/UIProcess/Media/RemoteMediaSessionProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::hasLocalStorageOrCookies):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::backForwardAddItemShared):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
(WebKit::WebFramePolicyListenerProxy::WebFramePolicyListenerProxy):
(WebKit::WebFramePolicyListenerProxy::didReceiveAppBoundDomainResult):
(WebKit::WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults):
(WebKit::WebFramePolicyListenerProxy::didReceiveInitialLinkDecorationFilteringData):
(WebKit::WebFramePolicyListenerProxy::didReceiveSiteHasStorageResults):
(WebKit::WebFramePolicyListenerProxy::use):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
(WebKit::WebFramePolicyListenerProxy::create):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::setUpPolicyListenerProxy):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForNewWindowAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::enhancedSecurityEnabled const):
(WebKit::WebPageProxy::beginSiteHasStorageCheck):
(WebKit::WebPageProxy::shouldEnableEnhancedSecurity const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::takeProcess):
(WebKit::WebProcessCache::takeSharedProcess):
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcessPool::createNewWebProcess):
(WebKit::WebProcessPool::tryTakePrewarmedProcess):
(WebKit::WebProcessPool::prewarmProcess):
(WebKit::WebProcessPool::processForSite):
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prepareProcessForNavigation):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultEnhancedSecurityDirectory):
(WebKit::WebsiteDataStore::enhancedSecuritySitesHolder):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp: Added.
(WebKit::EnhancedSecuritySitesHolder::sharedWorkQueueSingleton):
(WebKit::EnhancedSecuritySitesHolder::create):
(WebKit::EnhancedSecuritySitesHolder::EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::~EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::fetchEnhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesHolder::fetchAllEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesHolder::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesHolder::deleteSites):
(WebKit::EnhancedSecuritySitesHolder::deleteAllSites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h: Added.
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp: Added.
(WebKit::EnhancedSecuritySitesPersistence::EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::~EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::reportSQLError):
(WebKit::databasePath):
(WebKit::EnhancedSecuritySitesPersistence::checkedDatabase const):
(WebKit::EnhancedSecuritySitesPersistence::cachedStatement):
(WebKit::EnhancedSecuritySitesPersistence::openDatabase):
(WebKit::EnhancedSecuritySitesPersistence::deleteSite):
(WebKit::EnhancedSecuritySitesPersistence::deleteSites):
(WebKit::EnhancedSecuritySitesPersistence::deleteAllSites):
(WebKit::EnhancedSecuritySitesPersistence::fetchEnhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesPersistence::fetchAllEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesPersistence::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesPersistence::closeDatabase):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.h: Added.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::hasLocalStorageOrCookies const):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::enhancedSecurityDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setEnhancedSecurityDirectory):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.h:
* Source/WebKit/UIProcess/mac/WKTextFinderClient.mm:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showCaptionDisplaySettings):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm: Added.
(-[TestUIDelegate waitForAlertWithEnhancedSecurity]):
(-[WKWebView _test_waitForAlertWithEnhancedSecurity]):
(testAlertWithEnhancedSecurity):
(enhancedSecurityTestConfiguration):
(runActionAndCheckEnhancedSecurityAlerts):
(loadRequestAndCheckEnhancedSecurityAlerts):
(runHttpLoad):
(runHttpsLoad):
(runSameSiteHttpsUpgrade):
(runHttpEmbeddingHttpIframe):
(runHttpEmbedHttpsIframe):
(runCrossSiteHttpRedirect):
(runCrossSiteHttpToHttpsRedirect):
(runHttpOpeningHttpsWindow):
(runHttpOpeningHttpsTargetSelf):
(runHttpOpeningHttpsNoOpener):
(runHttpLocationRedirectsHttps):
(runHttpThenUserNavigateToHttps):
(runHttpThenClickLinkToHttps):
(runHttpsToHttpsThenBack):
(runHttpNavigateToHttpsThenBack):
(runMultiHopThenBack):
(runMultiHopThenBackJavascript):
(runMultiHopThenBackToSecure):
(runMultiHopThenBackToSecureJavascript):
(runHttpThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithLocalStorage):
(runHttpThenHttpsThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithCookiesThenHttps):
(runHttpThenNavigateToHttpsSetCookiesThenNavigateToHttpsAgain):
(runWindowOpenThenNavigateToMeaningfulSite):
(runReloadEnhancedSecurityRemains):
(runJavascriptRefreshEnhancedSecurityRemains):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPI):
(enhancedSecuritySitesPath):
(emptyEnhancedSecuritySitesPath):
(createEnhancedSecuritySitesTable):
(addEnhancedSecuritySite):
(setUpEnhancedSecuritySeenValues):
(cleanUpEnhancedSecuritySites):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAndExpectations):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndNotSeenOutsideEnhancedSecurity):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndSeenOutsideEnhancedSecurity):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageUtilities.mm:
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea4c174682b21a738dc9e29112659e7e121eba17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133518 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141074 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85569 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135388 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5884 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102132 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69524 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136465 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4627 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119692 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82931 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4506 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2109 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125590 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113632 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37804 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143721 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132029 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5689 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38388 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110512 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5771 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4877 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110694 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4366 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115950 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59439 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5744 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34262 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164994 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5591 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69196 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43116 "Found 4 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/regexp-zero-length-alternatives.js.layout-no-llint, microbenchmarks/regexp-prototype-search-complex-pattern.js.no-cjit-collect-continuously, stress/simple-regexp-test-folding-fail.js.dfg-eager, wasm.yaml/wasm/function-tests/memory-alignment.js.wasm-eager (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5833 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5700 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/UIProcess/Extensions/glib/WebExtensionGLib.cpp:
```diff
@@ -30,6 +30,7 @@
 #include <wtf/FileSystem.h>
 #include <wtf/Language.h>
 #include <wtf/glib/GRefPtr.h>
+#include <wtf/text/MakeString.h>
```
Needed?

ronan-apple replied:
I think so - this was a unified build fix.

sysrqb commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,237 @@
+    return staticEnabledSites;
+}
+
+static bool sitePreviouslyUsedEnhancedSecurity(API::Navigation& navigation)
```
Maybe `didSitePreviouslyUseEnhancedSecurity`

sysrqb replied:
or `wasSitePreviouslyUsed...`

ronan-apple replied:
Changed! I opted for `didSitePreviouslyUseEnhancedSecurity`.

sysrqb commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.h:
```diff
@@ -0,0 +1,81 @@
+
+enum class EnhancedSecurityReason : uint8_t { None, InsecureProvisional, InsecureLoad, Policy };
+
+typedef HashMap<WebCore::RegistrableDomain, EnhancedSecurityReason> EnhancedSecuritySites;
```
Prefer `using` over `typedef.

```suggestion
using EnhancedSecuritySites = HashMap<WebCore::RegistrableDomain, EnhancedSecurityReason>;
```

ronan-apple replied:
Ah, thank you, fixed!

Commit:
Implement EnhancedSecurity HTTP and heuristics tracking.
https://bugs.webkit.org/show_bug.cgi?id=302003
rdar://159219681

Reviewed by NOBODY (OOPS!).

This change enables usage of EnhancedSecurity when various conditions occur.

In particular, if an insecure HTTP request is made for the main frame, which
is not later upgraded to a HTTPS site, we will enable EnhancedSecurity.

Various heuristics then apply to determine whether we will retain this
EnhancedSecurity state for subsequent navigations, which this change implements.

When a UI initiated navigation occurs we will disable EnhancedSecurity, dropping
back to a regular tab state.

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::hasLocalStorageOrCookies):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/EnhancedSecurity.h: Added.
(WebKit::enhancedSecurityEnabledForState):
(WebKit::enhancedSecurityStatesAreConsistent):
* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::setEnhancedSecurity):
(WebKit::WebBackForwardListItem::enhancedSecurity const):
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
(WebKit::toString):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::setCurrentRequest):
(API::Navigation::setSiteHasStorage):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::currentRequestSiteHasStorage const):
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _allWebsiteDataTypesIncludingPrivate]):
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp: Added.
(WebKit::enabledSitesMap):
(WebKit::didSitePreviouslyUseEnhancedSecurity):
(WebKit::EnhancedSecurityTracking::initializeWithWebsiteDataStore):
(WebKit::EnhancedSecurityTracking::updateEnhancedSecurityDomains):
(WebKit::EnhancedSecurityTracking::transfer):
(WebKit::EnhancedSecurityTracking::enhancedSecurityState const):
(WebKit::EnhancedSecurityTracking::reset):
(WebKit::EnhancedSecurityTracking::makeDormant):
(WebKit::EnhancedSecurityTracking::makeActive):
(WebKit::reasonForEnhancedSecurity):
(WebKit::EnhancedSecurityTracking::enableFor):
(WebKit::EnhancedSecurityTracking::trackChangingSiteNavigation):
(WebKit::EnhancedSecurityTracking::trackSameSiteNavigation):
(WebKit::EnhancedSecurityTracking::enableIfRequired):
(WebKit::EnhancedSecurityTracking::handleBackForwardNavigation):
(WebKit::EnhancedSecurityTracking::trackNavigation):
* Source/WebKit/UIProcess/EnhancedSecurityTracking.h: Added.
* Source/WebKit/UIProcess/Media/RemoteMediaSessionManagerProxy.cpp:
* Source/WebKit/UIProcess/Media/RemoteMediaSessionProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::hasLocalStorageOrCookies):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::WebBackForwardList::backForwardAddItemShared):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.cpp:
(WebKit::WebFramePolicyListenerProxy::WebFramePolicyListenerProxy):
(WebKit::WebFramePolicyListenerProxy::didReceiveAppBoundDomainResult):
(WebKit::WebFramePolicyListenerProxy::didReceiveSafeBrowsingResults):
(WebKit::WebFramePolicyListenerProxy::didReceiveInitialLinkDecorationFilteringData):
(WebKit::WebFramePolicyListenerProxy::didReceiveSiteHasStorageResults):
(WebKit::WebFramePolicyListenerProxy::use):
* Source/WebKit/UIProcess/WebFramePolicyListenerProxy.h:
(WebKit::WebFramePolicyListenerProxy::create):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::setUpPolicyListenerProxy):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForNewWindowAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::enhancedSecurityEnabled const):
(WebKit::WebPageProxy::beginSiteHasStorageCheck):
(WebKit::WebPageProxy::shouldEnableEnhancedSecurity const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::takeProcess):
(WebKit::WebProcessCache::takeSharedProcess):
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcessPool::createNewWebProcess):
(WebKit::WebProcessPool::tryTakePrewarmedProcess):
(WebKit::WebProcessPool::prewarmProcess):
(WebKit::WebProcessPool::processForSite):
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prepareProcessForNavigation):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultEnhancedSecurityDirectory):
(WebKit::WebsiteDataStore::enhancedSecuritySitesHolder):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp: Added.
(WebKit::EnhancedSecuritySitesHolder::sharedWorkQueueSingleton):
(WebKit::EnhancedSecuritySitesHolder::create):
(WebKit::EnhancedSecuritySitesHolder::EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::~EnhancedSecuritySitesHolder):
(WebKit::EnhancedSecuritySitesHolder::fetchEnhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesHolder::fetchAllEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesHolder::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesHolder::deleteSites):
(WebKit::EnhancedSecuritySitesHolder::deleteAllSites):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.h: Added.
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp: Added.
(WebKit::EnhancedSecuritySitesPersistence::EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::~EnhancedSecuritySitesPersistence):
(WebKit::EnhancedSecuritySitesPersistence::reportSQLError):
(WebKit::databasePath):
(WebKit::EnhancedSecuritySitesPersistence::checkedDatabase const):
(WebKit::EnhancedSecuritySitesPersistence::cachedStatement):
(WebKit::EnhancedSecuritySitesPersistence::openDatabase):
(WebKit::EnhancedSecuritySitesPersistence::deleteSite):
(WebKit::EnhancedSecuritySitesPersistence::deleteSites):
(WebKit::EnhancedSecuritySitesPersistence::deleteAllSites):
(WebKit::EnhancedSecuritySitesPersistence::fetchEnhancedSecurityOnlyDomains):
(WebKit::EnhancedSecuritySitesPersistence::fetchAllEnhancedSecuritySites):
(WebKit::EnhancedSecuritySitesPersistence::trackEnhancedSecurityForDomain):
(WebKit::EnhancedSecuritySitesPersistence::closeDatabase):
* Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.h: Added.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::resolveDirectories):
(WebKit::WebsiteDataStore::fetchDataAndApply):
(WebKit::WebsiteDataStore::removeData):
(WebKit::WebsiteDataStore::hasLocalStorageOrCookies const):
(WebKit::WebsiteDataStore::removeEnhancedSecuritySites):
(WebKit::WebsiteDataStore::removeAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::fetchEnhancedSecurityOnlyDomains):
(WebKit::WebsiteDataStore::fetchAllEnhancedSecuritySites):
(WebKit::WebsiteDataStore::trackEnhancedSecurityForDomain):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy const):
(WebKit::WebsiteDataStoreConfiguration::Directories::isolatedCopy):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
(WebKit::WebsiteDataStoreConfiguration::enhancedSecurityDirectory const):
(WebKit::WebsiteDataStoreConfiguration::setEnhancedSecurityDirectory):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.h:
* Source/WebKit/UIProcess/mac/WKTextFinderClient.mm:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::showCaptionDisplaySettings):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurityPolicies.mm: Added.
(-[TestUIDelegate waitForAlertWithEnhancedSecurity]):
(-[WKWebView _test_waitForAlertWithEnhancedSecurity]):
(testAlertWithEnhancedSecurity):
(enhancedSecurityTestConfiguration):
(runActionAndCheckEnhancedSecurityAlerts):
(loadRequestAndCheckEnhancedSecurityAlerts):
(runHttpLoad):
(runHttpsLoad):
(runSameSiteHttpsUpgrade):
(runHttpEmbeddingHttpIframe):
(runHttpEmbedHttpsIframe):
(runCrossSiteHttpRedirect):
(runCrossSiteHttpToHttpsRedirect):
(runHttpOpeningHttpsWindow):
(runHttpOpeningHttpsTargetSelf):
(runHttpOpeningHttpsNoOpener):
(runHttpLocationRedirectsHttps):
(runHttpThenUserNavigateToHttps):
(runHttpThenClickLinkToHttps):
(runHttpsToHttpsThenBack):
(runHttpNavigateToHttpsThenBack):
(runMultiHopThenBack):
(runMultiHopThenBackJavascript):
(runMultiHopThenBackToSecure):
(runMultiHopThenBackToSecureJavascript):
(runHttpThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithLocalStorage):
(runHttpThenHttpsThenNavigateToHttpsSiteWithCookies):
(runHttpThenNavigateToHttpsSiteWithCookiesThenHttps):
(runHttpThenNavigateToHttpsSetCookiesThenNavigateToHttpsAgain):
(runWindowOpenThenNavigateToMeaningfulSite):
(runReloadEnhancedSecurityRemains):
(runJavascriptRefreshEnhancedSecurityRemains):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPI):
(enhancedSecuritySitesPath):
(emptyEnhancedSecuritySitesPath):
(createEnhancedSecuritySitesTable):
(addEnhancedSecuritySite):
(setUpEnhancedSecuritySeenValues):
(cleanUpEnhancedSecuritySites):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAndExpectations):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndNotSeenOutsideEnhancedSecurity):
(runHttpThenNavigateToHttpsSiteWithCookiesViaAPIAndSeenOutsideEnhancedSecurity):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageUtilities.mm:
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):

sysrqb commented on Source/WebKit/UIProcess/Media/RemoteMediaSessionProxy.h:
```diff
@@ -28,6 +28,8 @@
 #if ENABLE(VIDEO) || ENABLE(WEB_AUDIO)
 
+#include "RemoteMediaSessionClientProxy.h"
+#include "RemoteMediaSessionManagerProxy.h"
```
Unified build fix?

ronan-apple replied:
Yep!

sysrqb commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,243 @@
+        return reply();
+    }
+
+    hasMeaningfulSiteInteraction(navigation, [
```
Instead of calling this after we receive the policy response, we should prefer calling this in parallel. You can see an example of this in the safe browsing check beginning in `WebPageProxy::decidePolicyForNavigationAction` (`beginSafeBrowsingCheck`). `WebFramePolicyListenerProxy` will block the callback response until we get all of the responses that we need.

ronan-apple replied:
Great point, thank you for pointing this out and suggesting where to look for existing examples! I did exactly as suggested and put this in `WebFramePolicyListenerProxy` - so we now perform this check in parallel and block until we get all the responses back. This is only done if needed (e.g. if the heuristics are enabled), and appears to function well from my testing / automated tests.

sysrqb commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -86,7 +86,7 @@ constexpr auto countSubresourceUniqueRedirectsToQuery = "SELECT COUNT(*) FROM Su
 constexpr auto insertObservedDomainQuery = "INSERT INTO ObservedDomains (registrableDomain, lastSeen, hadUserInteraction,"
     "mostRecentUserInteractionTime, grandfathered, isPrevalent, isVeryPrevalent, dataRecordsRemoved, timesAccessedAsFirstPartyDueToUserInteraction,"
-    "timesAccessedAsFirstPartyDueToStorageAccessAPI, isScheduledForAllButCookieDataRemoval, mostRecentWebPushInteractionTime) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"_s;
+    "timesAccessedAsFirstPartyDueToStorageAccessAPI, isScheduledForAllButCookieDataRemoval, mostRecentWebPushInteractionTime, seenOutsideEnhancedSecurity) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"_s;
```
I agree this is a reasonable change in theory, but it's a little more complicated. ResourceLoadStatistics is tightly coupled with cross-site tracking prevention (i.e., ITP) such that we don't have this database if a customer disables tracking prevention. We can think about decoupling these and actually tracking some of these statistics separately from the tracking prevention, but that's a larger project.

ronan-apple replied:
Very valid - thanks for pointing this out! I've now adjusted this to use its own separate storage type, similar to recent searches. This means introducing a new WebsiteDataType which I've done so. I opted for a SQLiteDB instead of the plist that recent searches uses as I think this will be updated a little more frequently, and could grow enough in size that it warrants more efficient updates. This is wrapped and managed by `WebsiteData/EnhancedSecuritySitesHolder.cpp` and `WebsiteData/EnhancedSecuritySitesPersistence.cpp`.

sysrqb commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2954,4 +2959,14 @@ void WebsiteDataStore::isStorageSuspendedForTesting(CompletionHandler<void(bool)
 
+void WebsiteDataStore::setSeenOutsideEnhancedSecurity(const URL& url) const
+{
+    protectedNetworkProcess()->setSeenOutsideEnhancedSecurity(m_sessionID, WebCore::RegistrableDomain { url });
```
Instead, maybe explore how we are tracking recent searches (`WebsiteDataStore::{save,load,remove}RecentSearches`)

ronan-apple replied:
Thanks for pointing these examples out! These were really helpful to be able to go and see how they work.

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,300 @@
+{
+    ASSERT(!isMainRunLoop());
+
+    FileSystem::makeAllDirectories(directoryPath);
```
This should be called if `directoryPath` is not empty.
Also for ephemeral website data store, is there a need to create a database at all? Don't we have an in-memory map?

ronan-apple replied:
Excellent point - yes, we have no need to create the database at all and, in fact, shouldn't use `EnhancedSecuritySitesPersistence` if we're an ephemeral website data store. I've added checks in `WebsiteDataStoreCocoa` that return early, and an `ASSERT` on the initialisation path for `EnhancedSecuritySitesHolder` to ensure this is the case.

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,300 @@
+    for (auto& site : sites)
+        deleteSite(site);
+
+    callOnMainRunLoop(WTFMove(completionHandler));
```
The `completionHandler` should be called on the same thread / runloop that the function runs, i.e. it will be the caller of `deleteSites` to dispatch the callback to main runloop.
Same for the other functions.

ronan-apple replied:
Ah, right, thank you! I've fixed this, and added a wrapper completion handler that does the main runloop dispatch in `EnhancedSecuritySitesHolder` instead.

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,300 @@
+static constexpr auto enhancedSecurityStateIndexName = "idx_sites_enhanced_security_state"_s;
+
+static constexpr auto createSitesTableSQL = "CREATE TABLE sites (site TEXT PRIMARY KEY NOT NULL, enhanced_security_state INT NOT NULL)"_s;
+static constexpr auto createSiteIndexSQL =  "CREATE INDEX idx_sites_site ON sites(site)"_s;
```
Since site is the primary key in the table, there is no need to create a separate index for it.

ronan-apple replied:
Oh, yes, great point - thank you - removed!

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,300 @@
+#if RELEASE_LOG_DISABLED
+    UNUSED_PARAM(method);
+    UNUSED_PARAM(action);
+#endif
```
Not needed

ronan-apple replied:
I've removed this but didn't quite follow why this wasn't needed? Is `RELEASE_LOG_DISABLED` not a thing that can occur anymore?

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesPersistence.cpp:
```diff
@@ -0,0 +1,300 @@
+        || selectSiteStatement->bindText(1, site.string()) != SQLITE_OK)
+        reportSQLError("trackEnhancedSecurityForDomain"_s, "Failed to query specific site"_s);
+
+    if (selectSiteStatement->step() == SQLITE_ROW) {
+        if (static_cast<EnhancedSecurity>(selectSiteStatement->columnInt(0)) == EnhancedSecurity::Disabled)
+            reason = EnhancedSecurity::Disabled;
+    }
```
So if the stored reason is disabled we will use disabled; then why not just early return here?

ronan-apple replied:
Excellent point, I think I confused myself implementing this! Updated to just return early in this case.

szewai commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -80,6 +80,8 @@
 static constexpr double defaultAppTotalQuotaRatio = 0.2;
 static constexpr double defaultAppOriginQuotaRatio = 0.15;
 
+static constexpr Seconds secondsToWaitForDispatchEnhancedSecuritySitesSave = 3_s;
```
Unused

ronan-apple replied:
Oops, great point, removed!

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,141 @@
+    sharedWorkQueueSingleton()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, completionHandler = WTFMove(completionHandler)] mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
```
`completionHandler` is not called

ronan-apple replied:
Ah, yep, fixed!

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,141 @@
+    sharedWorkQueueSingleton()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, completionHandler = WTFMove(completionHandler)] mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis)
+            return;
```
`completionHandler` is not called

ronan-apple replied:
Ah, yep, fixed!

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,141 @@
+    if (domain.isEmpty())
+        return;
+
+    sharedWorkQueueSingleton()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, domain = crossThreadCopy(domain), reason]() mutable {
```
crossThreadCopy(WTFMove(domain))

ronan-apple replied:
Ah yep, thank you, addressed!

szewai commented on Source/WebKit/UIProcess/WebsiteData/EnhancedSecuritySitesHolder.cpp:
```diff
@@ -0,0 +1,141 @@
+    });
+}
+
+void EnhancedSecuritySitesHolder::deleteSites(Vector<WebCore::RegistrableDomain>& sites, CompletionHandler<void()>&& completionHandler)
```
Either `Vector<WebCore::RegistrableDomain>&&` or `const Vector<WebCore::RegistrableDomain>&`

ronan-apple replied:
Updated to be `const Vector<WebCore::RegistrableDomain>&` and propagated that through - thank you!

szewai reviewed (commented):
Only checking the storage related code

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1163,6 +1163,26 @@ void NetworkProcess::clearUserInteraction(PAL::SessionID sessionID, RegistrableD
+    if (!session || !networkStorageSession)
+        return completionHandler(false);
+
+    networkStorageSession->hasCookies(domain, [session, domain, completionHandler = WTFMove(completionHandler)](bool hasCookies) mutable {
```
Capturing CheckedPtr in asynchronous lambdas is unusual and risky given that they crash if the object gets destroyed. We usually capture weak pointers and convert them back into stronger pointers in the lambda.

ronan-apple replied:
Ah, great point, thanks for the explanation - fixed!

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1163,6 +1163,26 @@ void NetworkProcess::clearUserInteraction(PAL::SessionID sessionID, RegistrableD
+        if (hasCookies)
+            return completionHandler(true);
+
+        session->storageManager().fetchData({ WebsiteDataType::LocalStorage }, NetworkStorageManager::ShouldComputeSize::No, [domain, completionHandler = WTFMove(completionHandler)](auto entries) mutable {
```
`domain = WTFMove(domain)`

ronan-apple replied:
Ah, thanks, fixed!

cdumez commented on Source/WebKit/Shared/EnhancedSecurity.h:
```diff
@@ -0,0 +1,47 @@
+enum class EnhancedSecurity : uint8_t { Disabled, EnabledInsecure, EnabledPolicy };
+enum class EnhancedSecurityReason : uint8_t { None, InsecureProvisional, InsecureLoad, Policy };
+
+using EnhancedSecuritySites = HashMap<WebCore::RegistrableDomain, EnhancedSecurityReason>;
```
Can we include `Map` in the name for clarity?

ronan-apple replied:
Ah, yes, good point - I moved this to a more appropriate place, and renamed to include Map in the name.

cdumez commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -208,4 +209,11 @@ WTF::String Navigation::loggingString() const
+
+void Navigation::setSiteHasStorage(const WTF::URL& url, bool siteHasStorage)
+{
+    ASSERT(url == m_currentRequest.url());
```
`ASSERT_UNUSED(url, url == m_currentRequest.url());`

ronan-apple replied:
Ah, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -199,6 +199,9 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
     void setPendingSharedProcess(WebKit::FrameProcess&);
 
+    void setSiteHasStorage(const WTF::URL&, bool);
+    bool currentRequestSiteHasStorage() const { return m_siteHasStorage; }
```
The getter and setter should be consistent in terms of naming. Ditto for the data member.

ronan-apple replied:
Thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/API/APINavigation.h:
```diff
@@ -229,6 +232,7 @@ class Navigation : public ObjectImpl<Object::Type::Navigation> {
     bool m_requestIsFromClientInput : 1 { false };
     bool m_isFromLoadData : 1 { false };
     bool m_safeBrowsingCheckTimedOut : 1 { false };
+    bool m_siteHasStorage : 1 { false };
```
Per WebKit coding style, boolean variables need a prefix. e.g. `m_hasStorageForSite`

ronan-apple replied:
Thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
```diff
@@ -113,8 +113,9 @@ class WebsitePolicies final : public API::ObjectImpl<API::Object::Type::WebsiteP
 
-    bool enhancedSecurityEnabled() const { return m_enhancedSecurityEnabled; }
-    void setEnhancedSecurityEnabled(bool enabled) { m_enhancedSecurityEnabled = enabled; }
+    bool enhancedSecurityEnabled() const { return m_enhancedSecurityEnabled.value_or(false); }
```
We usually name these things `isEnhancedSecurityEnabled()`, with a prefix since this is a boolean.
In the API folder, we sometimes opts with a verb also such as `enablesEnhancedSecurity()` though I personally prefer the former.

ronan-apple replied:
Yep, this makes sense - thanks for flagging, have fixed this name.

cdumez commented on Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
```diff
@@ -173,7 +174,7 @@ class WebsitePolicies final : public API::ObjectImpl<API::Object::Type::WebsiteP
     RefPtr<WebKit::WebUserContentControllerProxy> m_userContentController;
     std::optional<bool> m_lockdownModeEnabled;
-    bool m_enhancedSecurityEnabled { false };
+    std::optional<bool> m_enhancedSecurityEnabled;
```
Missing prefix.

ronan-apple replied:
Thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+static EnhancedSecuritySites& enabledSitesMap()
+{
+    ASSERT(isMainRunLoop());
+    static NeverDestroyed<EnhancedSecuritySites> staticEnabledSites;
```
You can use `MainThreadNeverDestroyed` here and omit the assertion above.

ronan-apple replied:
Oh that's lovely - great to know, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    return staticEnabledSites;
+}
+
+static bool didSitePreviouslyUseEnhancedSecurity(API::Navigation& navigation)
```
Can we make the parameter `const`?

ronan-apple replied:
Yep, good point, fixed!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+
+void EnhancedSecurityTracking::initializeWithWebsiteDataStore(Ref<WebsiteDataStore> websiteDataStore)
+{
+    websiteDataStore->fetchEnhancedSecurityOnlyDomains([this](HashSet<RegistrableDomain>&& domains) {
```
Capture `this` like this goes against our safer cpp guidelines.

ronan-apple replied:
I struggled to change this, I was hoping to get your suggestion on how to sort it. Do I need to change EnhancedSecurityTracking to inherit from CanMakeWeakPtr? RefCountedAndCanMakeWeakPtr? Or is there a better way to fix this with the single way we use EnhancedSecurityTracking?

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    return enabledSitesMap().contains(RegistrableDomain { navigation.currentRequest().url() });
+}
+
+void EnhancedSecurityTracking::initializeWithWebsiteDataStore(Ref<WebsiteDataStore> websiteDataStore)
```
The parameter should not be a smart pointer here and certainly not passed by value. I think we would normally pass a `WebsiteDataStore&` here.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+
+void EnhancedSecurityTracking::updateEnhancedSecurityDomains(HashSet<RegistrableDomain>&& domains)
+{
+    for (auto domain : domains)
```
`auto&` to avoid ref-counting churn.

ronan-apple replied:
Ah, thank you, fixed.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+void EnhancedSecurityTracking::updateEnhancedSecurityDomains(HashSet<RegistrableDomain>&& domains)
+{
+    for (auto domain : domains)
+        enabledSitesMap().add(domain, EnhancedSecurityReason::InsecureLoad);
```
Question: What do you want to happen if the domain is already in the map? Should we overwrite the value with `InsecureLoad`? If so, you want to call `set()`. If you call `add()` this will be a no-op if the domain is already in the map. I'm not sure what is the expected behavior but I wanted to make sure you considered this.

ronan-apple replied:
Great point - I think `add` here is appropriate, it would be best to take the already existing one in the map if it did, but I've applied the correct thinking elsewhere and changed to `set`.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+        enabledSitesMap().add(domain, EnhancedSecurityReason::InsecureLoad);
+}
+
+void EnhancedSecurityTracking::transfer(EnhancedSecurityTracking& other)
```
Can we make the parameter `const`?

ronan-apple replied:
Yep, applied!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    ASSERT_NOT_REACHED();
+}
+
+void EnhancedSecurityTracking::enableFor(EnhancedSecurityReason reason, API::Navigation& navigation)
```
Can the navigation parameter be `const`?

ronan-apple replied:
Yep, applied!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    m_activeReason = reason;
+    m_initialProtectedDomain = RegistrableDomain(navigation.currentRequest().url());
+
+    enabledSitesMap().add(m_initialProtectedDomain, m_activeReason);
```
Please consider whether `add()` or `set()` is the behavior you want here.

ronan-apple replied:
Yep, great point, I think we'd want `set` here instead.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    enabledSitesMap().add(m_initialProtectedDomain, m_activeReason);
+}
+
+void EnhancedSecurityTracking::trackChangingSiteNavigation(API::Navigation& navigation)
```
How come the parameter is not used?

ronan-apple replied:
Ah great point, I've actually removed this parameter in the new PR - we already have the meaningful knowledge that this is a changing site navigation, which is all we needed to apply the logic in the function.

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    if (enhancedSecurityReason() == EnhancedSecurityReason::InsecureProvisional) {
+        m_activeReason = EnhancedSecurityReason::InsecureLoad;
+
+        if (enabledSitesMap().contains(m_initialProtectedDomain)) {
```
Please use `HashMap::find()` to go from 3 HashMap look-ups to a single one.

ronan-apple replied:
This is lovely - great point, thank you - fixed!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    }
+}
+
+void EnhancedSecurityTracking::trackSameSiteNavigation(API::Navigation& navigation)
```
Can the parameter be `const`?

ronan-apple replied:
Yep, fixed!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+
+void EnhancedSecurityTracking::trackSameSiteNavigation(API::Navigation& navigation)
+{
+    bool isHttps = navigation.currentRequest().url().protocolIs("https"_s);
```
I think `isHTTPS` would be more common in WebKit style since HTTP is an abbreviation.

ronan-apple replied:
Ah, good point, thanks, fixed!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+    if (enhancedSecurityReason() == EnhancedSecurityReason::InsecureProvisional && isHttps) {
+        reset();
+
+        auto domain = RegistrableDomain { navigation.currentRequest().url() };
```
` RegistrableDomain domain { navigation.currentRequest().url() };` is more common and more concise.

ronan-apple replied:
Oh neat, yep, this is much better, done!

cdumez commented on Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp:
```diff
@@ -0,0 +1,234 @@
+        reset();
+
+        auto domain = RegistrableDomain { navigation.currentRequest().url() };
+        if (enabledSitesMap().contains(domain)) {
```
Please use `HashMap::find()` to go from 3 HashMap lookups to a single one.

ronan-apple replied:
Ah yes, perfect, thank you - this is really nice, fixed!

cdumez reviewed (commented):
I did not finish reviewing. This patch is too large. Is there a way we can split this PR into multiple ones (this is the preferred practice in WebKit), so they can get properly reviewed. The PR is so large even typing review comments is slow in Safari for me.
